### ISSUE TITLE
Review Authentication Request Controller Logic

### ISSUE THREAD

**Issue by rwinch** [spring-projects/spring-security-saml#351]
* The AuthenticationEntryPoint should be configured with a URL that aligns better with the protocol. Perhaps something like `/saml2/authentication/request/{idp-id}`
* The service provider should process the URL from above and then send an authentication request to the idp that has an id of `idp-id`
* SamlProcessingFilter should probably be removed. 
  * All processing of the SAMLResponse should be done in the authentication filter
  * I'm not sure what the SAMLRequest is doing in this filter since this is something that would be processed by the IdP and not the SP? 
* SelectIdentityProviderUIFilter 
  * should be renamed to align with the specification Something like Saml2AuthenticationRequestRedirectFilter
  * Rather than having a dependency on any view technology or having to abstract out the view technology, it should just support forwarding to a URL with the model (on the request attributes set). This allows an application to use whatever technology it wants for the UI. 
  * The model should also be something like ServiceProviderRegistration (which would align with the ClientRegistration)
  * If no forward URL is configured it should just perform the redirect

**Comment by fhanik** [comment]
@rwinch Thanks for the feedback. Let me gather a bit more information so I can fully understand the intent, and also highlight some of the implications.

> The AuthenticationEntryPoint should be configured with a URL that aligns better with the protocol. Perhaps something like /saml2/authentication/request/{idp-id}

That is easily done. However, it would break the external HTTP API for those that are upgrading. We're already forcing developers to breaking code upgrade internally inside their application. With this change, we'd force all external parties to that application to deal with a breaking URL/API upgrade.

This means, that there is no way for me to replace Spring Security SAML v1 with v3 or later because I would have to go and reconfigure any external entity too. That's probably a deal breaker for anyone wanting to use SAML while trying to upgrade an existing environment

It would be very easy to support both. Since we spoke of backwards compatibility yesterday, and how disciplined the entire framework is, maybe we should think twice before we break all the existing URLs?

>The service provider should process the URL from above and then send an authentication request to the idp that has an id of idp-id

Yes, that's fine. As long as a developer can override that behavior to maintain compatibility in existing systems. The SP initiated login functionality is pretty straight forward, but most systems will have custom URLs and not rely on our defaults ones. I just figured that rather than inventing a new one, give them one they may already have.

> SamlProcessingFilter should probably be removed.
All processing of the SAMLResponse should be done in the authentication filter
I'm not sure what the SAMLRequest is doing in this filter since this is something that would be processed by the IdP and not the SP?

Both SPs and IDPs are sending `SAMLRequest` objects and receiving `SAMLResponse` objects. Both parameters are valid for both types of entities. And they are named the same in all types of communication, `SAMLRequest` always is a request, regardless if the sender is an IDP or an SP. `SAMLResponse` is always a response, regardless if the sender is an IDP or an SP

This is standard SAML protocol. These parameter names are not reserved for a single type of message, but rather used globally across all types of SAML communications. 

>All processing of the SAMLResponse should be done in the authentication filter

That would mean a lot of redundant processing, after all, SamlProcessingFilter only does decoding and validity check. There is no need to continue down the filter chain if the decoding fails. It's a bad request.

So if we push this down to each SAML endpoint, it also means that each filter that processes any type of message would have to account for invalid message data (parse errors, encoding errors, etc), and each filter would have to return 400 Bad Request, when all of that can be trapped at once. This would be a performance decrease and complexity increase that seems unjustified in this scenario. 

>SelectIdentityProviderUIFilter
should be renamed to align with the specification Something like Saml2AuthenticationRequestRedirectFilter

That's an easy change. 

>Rather than having a dependency on any view technology or having to abstract out the view technology, it should just support forwarding to a URL with the model

We need a minimum of one way to send down HTML/Javascript to a user-agent in order to support the `POST` binding in SAML. Let me digest this proposition a bit as we need one default implementation for POST bindings, and writing a custom template engine seems redundant when the external library already includes one.

>The model should also be something like ServiceProviderRegistration (which would align with the ClientRegistration)

Can you clarify this? In OAuth2, a client is an external entity, an application that acts on behalf of a user (or itself, client_credentials). A client is registered with an authorization server. 

If we were to align it with OAuth2, when an app is acting as a ServiceProvider, it would actually have a collection of external `IdentityProviderRegistration` objects. Likewise, an IdentityProvider would have a collection of external ServiceProviderRegistration. So it seems unnatural that a hosted service provider would be a _registration_ in this case. 

The model has to reflect that, for a Service Provider,
1. A system can have one or more Service Providers hosted locally
2. Each service provider can have one or more external Identity Providers registered with it

Bearing that in mind, what is a _ServiceProviderRegistration_ exactly? That's what I'm having a hard time wrapping my brain around.

>If no forward URL is configured it should just perform the redirect

There are multiple IDPs to choose from, so just performing a redirect to an arbitrary IDP would not seem intuitive. And in the case of the remote IDP not supporting the REDIRECT binding, it would still have to be a HTML/Javascript POST via the browser. So the type of response is dependent on the external entity that is being responded to.

**Comment by rwinch** [comment]
> It would be very easy to support both. Since we spoke of backwards compatibility yesterday, and how disciplined the entire framework is, maybe we should think twice before we break all the existing URLs?

I think as long as we support both that we want consistency over backward compatibility. We already know things won't be backward compatible. As long as users have a path to preserve the HTTP APIs, I don't see a problem with changing default URLs either.

> This is standard SAML protocol. These parameter names are not reserved for a single type of message, but rather used globally across all types of SAML communications.

What is the SAMLRequest being used for by the SP? The logic belongs in that Filter.

> That would mean a lot of redundant processing, after all, SamlProcessingFilter only does decoding and validity check. There is no need to continue down the filter chain if the decoding fails. It's a bad request.
>
> So if we push this down to each SAML endpoint, it also means that each filter that processes any type of message would have to account for invalid message data (parse errors, encoding errors, etc), and each filter would have to return 400 Bad Request, when all of that can be trapped at once. This would be a performance decrease and complexity increase that seems unjustified in this scenario.

This can be encapsulated in other ways. Having a Filter that sets a request attribute does not follow the Spring Security way.

> Bearing that in mind, what is a ServiceProviderRegistration exactly? That's what I'm having a hard time wrapping my brain around.

In OAuth, we must map endpoints for the Authorization Server and that Authorization Server has a client id / secret. In SAML we must map the endpoint for the IdP, the private key used by the SP, and the public key used by the IdP. This is what the ServiceProviderRegistration would entail. If there are more than one IdP that can be used there are more than one ServiceProviderRegistration that are registered.

> There are multiple IDPs to choose from, so just performing a redirect to an arbitrary IDP would not seem intuitive. And in the case of the remote IDP not supporting the REDIRECT binding, it would still have to be a HTML/Javascript POST via the browser. So the type of response is dependent on the external entity that is being responded to.

This filter would require something in the URL (an id) to determine which IdP to redirect to. If there are multiple IdP's, then the log in page would be in charge of rendering something to select which IdP (and thus a different URL to the Saml2AuthenticationRequestRedirectFilter. We could have the Filter have a default POST support too, but for now redirect seemed the easiest to support since it has no UI. Any customization to it would require a forward (not a redirect) to a URL that the user must provide arbitrary logic.

**Comment by fhanik** [comment]
> I think as long as we support both that we want consistency over backward compatibility. We already know things won't be backward compatible. As long as users have a path to preserve the HTTP APIs, I don't see a problem with changing default URLs either.

Let me rephrase that. Everything that hits the wire is backwards compatible today, and intentionally so. Only the code API and configuration changes. Changing URLs would be an intentional breaking point for external systems and that seems counter intuitive for adoption.

>What is the SAMLRequest being used for by the SP? The logic belongs in that Filter.

Simplest example would be
SPs and IDPs can send and receive LogoutRequest
SPs and IDPs can send and receive LogoutResponse

these are transported in parameters named `SAMLRequest` and `SAMLResponse` and go both ways.
But the protocol has others.

Unlike OAuth, where the Authorization Server exclusively receives requests and responds accordingly, SAML has two way communication. Each entity can initiate a request to the other.

>This can be encapsulated in other ways. Having a Filter that sets a request attribute does not follow the Spring Security way.

Can you elaborate on _Spring Security way_ and what you have in mind please?

I'm not a fan of ThreadLocals, which is prevalent in Spring Security, but I can certainly do that. Is there another way you had in mind?
The original implementation only had one filter, and delegated to handlers, where the handlers received data by method signatures. It was clean and easy to read, but was also recommended away from.

So, for now, can you be a bit more clear about what you have in mind so I know where you wish this to go?

>This is what the ServiceProviderRegistration would entail. If there are more than one IdP that can be used there are more than one ServiceProviderRegistration that are registered.

I think that naming would cause confusions, I'm perceiving you saying that `ServiceProviderRegistration` represents an IDP.

I think what you mean is that the name for that would be `IdentityProviderRegistration`, because it represents an Identity Provider.
Just like `ClientRegistration` represents a Client. (if I used your ServiceProviderRegistration analogy, you would name each client `AuthorizationServerRegistration` and there would be one for each client)

**Comment by rwinch** [comment]
> Simplest example would be
> SPs and IDPs can send and receive LogoutRequest
> SPs and IDPs can send and receive LogoutResponse

This would typically be done then in the corresponding Filter. If the logic needs to be shared, then an API could potentially be created for this. The Filter is a controller, so short of obtaining the parameter from the request and passing it to an API it shouldn't do much of anything.

> Can you elaborate on Spring Security way and what you have in mind please?

Do not process it and place it in a request attribute. Instead, don't process it until you are in a Filter that needs it. The reason for this is it becomes difficult for users to connect all the dots on what is happening. If they want to see how log out is handled it should be the log out filter. They shouldn't have to see how it is parsed from yet another Filter.

> I think that naming would cause confusions, I'm perceiving you saying that ServiceProviderRegistration represents an IDP.

It represents a mapping of a SP to a IdP.

**Comment by fhanik** [comment]
Feedback Completed

> AuthenticationEntryPoint should be configured with a URL that aligns better with the protocol.

authentication request filter will respond on <prefix>/authenticate By
default it is `/saml/sp/authenticate`
This initiates an authentication Request to be *sent* to the IDP in order to receive a Response 

> SamlProcessingFilter should probably be removed.

Removed

> SelectIdentityProviderUIFilter renamed to Saml2AuthenticationRequestRedirectFilter
That would be incorrect. the AuthenticationRequestFilter is a different endpoint, fulfilling a different task

> remove dependency on any view technology

Hardcoded HTML has been provided. SamlTemplateEngine removed

> ServiceProviderRegistration

While we could have `IdentityProviderRegistration` instead of `ExternalIdentityProviderConfiguration`, it was later mentioned that names are not a priority right now
